File: api/src/main/java/org/pmiops/workbench/audit/adapters/ProfileAuditAdapterServiceImpl.java
package org.pmiops.workbench.audit.adapters;

import java.time.Clock;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.inject.Provider;
import org.pmiops.workbench.audit.ActionAuditEvent;
import org.pmiops.workbench.audit.ActionAuditService;
import org.pmiops.workbench.audit.ActionType;
import org.pmiops.workbench.audit.AgentType;
import org.pmiops.workbench.audit.TargetType;
import org.pmiops.workbench.db.model.User;
import org.pmiops.workbench.model.Profile;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class ProfileAuditAdapterServiceImpl implements ProfileAuditAdapterService {

  private static final Logger logger =
      Logger.getLogger(ProfileAuditAdapterServiceImpl.class.getName());

  private final Provider<User> userProvider;
  private final ActionAuditService actionAuditService;
  private final Clock clock;

  @Autowired
  public ProfileAuditAdapterServiceImpl(
      Provider<User> userProvider, ActionAuditService actionAuditService, Clock clock) {
    this.userProvider = userProvider;
    this.actionAuditService = actionAuditService;
    this.clock = clock;
  }

  @Override
  public void fireCreateAction(Profile createdProfile) {}

  @Override
  public void fireUpdateAction(Profile previousProfile, Profile updatedProfile) {}

  // Each user is assumed to have only one profile, but we can't rely on
  // the userProvider if the user is deleted before the profile.
  @Override
  public void fireDeleteAction(long userId, String userEmail) {
    try {
      final ActionAuditEvent deleteProfileEvent =
          new ActionAuditEvent(
              clock.millis(),
              AgentType.USER,
              userId,
              userEmail,
              ActionAuditEvent.newActionId(),
              ActionType.DELETE,
              TargetType.PROFILE,
              null,
              userId,
              null,
              null);
      actionAuditService.send(deleteProfileEvent);
    } catch (RuntimeException e) {
      logAndSwallow(e);
    }
  }

  private void logAndSwallow(RuntimeException e) {
    logger.log(Level.WARNING, e, () -> "Exception encountered during audit.");
  }
}

File: api/src/main/java/org/pmiops/workbench/actionaudit/adapters/ProfileAuditAdapterServiceImpl.kt
package org.pmiops.workbench.actionaudit.adapters

import java.time.Clock
import java.util.logging.Level
import java.util.logging.Logger
import javax.inject.Provider
import org.pmiops.workbench.actionaudit.ActionAuditEvent
import org.pmiops.workbench.actionaudit.ActionAuditService
import org.pmiops.workbench.actionaudit.ActionType
import org.pmiops.workbench.actionaudit.AgentType
import org.pmiops.workbench.actionaudit.TargetType
import org.pmiops.workbench.actionaudit.targetproperties.ProfileTargetProperty
import org.pmiops.workbench.actionaudit.targetproperties.TargetPropertyExtractor
import org.pmiops.workbench.db.model.DbUser
import org.pmiops.workbench.model.Profile
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service

@Service
class ProfileAuditAdapterServiceImpl @Autowired
constructor(
    private val userProvider: Provider<DbUser>,
    private val actionAuditService: ActionAuditService,
    private val clock: Clock,
    @Qualifier("ACTION_ID") private val actionIdProvider: Provider<String>
) : ProfileAuditAdapterService {

    override fun fireCreateAction(createdProfile: Profile) {
        try {
            val propertiesByName: Map<String, String?> = TargetPropertyExtractor
                    .getPropertyValuesByName(ProfileTargetProperty.values(), createdProfile)
            val createEvents = propertiesByName.entries
                    .map {
                        ActionAuditEvent(
                            timestamp = clock.millis(),
                            actionId = actionIdProvider.get(),
                            actionType = ActionType.CREATE,
                            agentType = AgentType.USER,
                            agentId = userProvider.get().userId,
                            agentEmailMaybe = userProvider.get().email,
                            targetType = TargetType.PROFILE,
                            targetIdMaybe = userProvider.get().userId,
                            targetPropertyMaybe = it.key,
                            previousValueMaybe = null,
                            newValueMaybe = it.value)
                    }
            actionAuditService.send(createEvents)
        } catch (e: RuntimeException) {
            logAndSwallow(e)
        }
    }

    override fun fireUpdateAction(previousProfile: Profile, updatedProfile: Profile) {
        try {
            // determine changed fields
            val changesByProperty = TargetPropertyExtractor.getChangedValuesByName(
                    ProfileTargetProperty.values(),
                    previousProfile,
                    updatedProfile)
            val events = changesByProperty.entries
                    .map {
                        ActionAuditEvent(
                                timestamp = clock.millis(),
                                actionId = actionIdProvider.get(),
                                actionType = ActionType.EDIT,
                                agentType = AgentType.USER,
                                agentId = userProvider.get().userId,
                                agentEmailMaybe = userProvider.get().email,
                                targetType = TargetType.PROFILE,
                                targetIdMaybe = userProvider.get().userId,
                                targetPropertyMaybe = it.key,
                                previousValueMaybe = it.value.previousValue,
                                newValueMaybe = it.value.newValue)
                    }
            actionAuditService.send(events)
        } catch (e: java.lang.RuntimeException) {
            logAndSwallow(e)
        }
    }

    // Each user is assumed to have only one profile, but we can't rely on
    // the userProvider if the user is deleted before the profile.
    override fun fireDeleteAction(userId: Long, userEmail: String) {
        try {
            val deleteProfileEvent = ActionAuditEvent(
                    timestamp = clock.millis(),
                    actionId = actionIdProvider.get(),
                    actionType = ActionType.DELETE,
                    agentType = AgentType.USER,
                    agentId = userId,
                    agentEmailMaybe = userEmail,
                    targetType = TargetType.PROFILE,
                    targetIdMaybe = userId,
                    targetPropertyMaybe = null,
                    previousValueMaybe = null,
                    newValueMaybe = null
                    )
            actionAuditService.send(deleteProfileEvent)
        } catch (e: RuntimeException) {
            logAndSwallow(e)
        }
    }

    private fun logAndSwallow(e: RuntimeException) {
        logger.log(Level.WARNING, e) { "Exception encountered during audit." }
    }

    companion object {

        private val logger = Logger.getLogger(ProfileAuditAdapterServiceImpl::class.java.name)
    }
}
